persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -306,101 +309,122 @@ static class PossibleExpirationNotification {
+                  batchSize--;
+                  String keyStr = rs.getString(2);
+                  long expiryTime = rs.getLong(3);
+                  InputStream inputStream = rs.getBinaryStream(1);
+                  MarshalledValue value = JdbcUtil.unmarshall(inputStream, this.marshaller);
+                  result.add(new PossibleExpirationNotification(expiryTime, keyStr, value));
                }
-               unicastProcessor.onError(e);
-            } finally {
-               JdbcUtil.safeClose(rs);
-               JdbcUtil.safeClose(ps);
-               connectionFactory.releaseConnection(conn);
+               return Collections.unmodifiableList(result);
             }
-         }, "jdbcstringstore-purge");
-         return unicastProcessor;
-      });
+         } finally {
+            conn.commit();
+            conn.setAutoCommit(true);
+         }
+      }
    }
 
-   private long runBatchAndNotify(List<PossibleExpirationNotification> possible, PreparedStatement batchDelete,
-         FlowableProcessor<MarshallableEntry<K, V>> flowable) throws SQLException {
-      long purgeAmount = 0;
-      int[] results = batchDelete.executeBatch();
-      if (possible != null) {
-         for (int i = 0; i < results.length; ++i) {
-            PossibleExpirationNotification notification = possible.get(i);
-            if (results[i] != Statement.EXECUTE_FAILED) {
-               Object key = ((TwoWayKey2StringMapper) key2StringMapper).getKeyMapping(notification.key);
-               flowable.onNext(marshalledEntryFactory.create(key, notification.is));
-
-               purgeAmount++;
-            } else {
-               log.tracef("Unable to remove expired entry for key %s, most likely concurrent update", notification.key);
+   private int[] deleteEntries(List<PossibleExpirationNotification> result) throws SQLException {
+      int[] deleteResults;
+      try (var conn = this.connectionFactory.getConnection()) {
+         conn.setAutoCommit(false);
+         try (PreparedStatement batchDelete = conn.prepareStatement(this.getTableManager().getDeleteRowWithExpirationSql())) {
+            for (PossibleExpirationNotification notification : result) {
+               batchDelete.setString(1, notification.key);
+               batchDelete.setLong(2, notification.expiryTime);
+               batchDelete.addBatch();
+            }
+            deleteResults = batchDelete.executeBatch();
+            if (log.isTraceEnabled()) {
+               int purgedAmount = Arrays.stream(deleteResults).sequential().filter(r -> r != Statement.EXECUTE_FAILED).sum();
+               log.tracef("Successfully purged %d rows.", purgedAmount);
+            }
+         } catch (Exception ex) {
+            try {
+               conn.rollback();
+            } catch (SQLException rollbackException) {
+               log.sqlFailureTxRollback(rollbackException);
             }
+            throw ex;
+         } finally {
+            conn.commit();
```

Comment:
Same here, the commit should be the last line in the try block instead.

Reply:
Kept autocommit handling at deletion because of error by calling commit during autocommit == true. I am not sure how autocommit and batch-delete work together, I would prefer to keep using transactions here even in autocommit case. Transactions in case of autocommit == false are necessary anyway.